### ISSUE TITLE
[Suggestion] unnecessary late lint

### ISSUE THREAD

**Issue by bernaferrari** [dart-lang/linter#2730]
Correct code:
```dart
final String value;
if (a == b) {
    value = "hello";
} else {
    value = "world";
}
print(value);
```

Not wrong code:
```dart
late final String value; // I would like a warning here to remove the late.
if (a == b) {
    value = "hello";
} else {
    value = "world";
}
print(value);
```

Wrong code:
```dart
final String value;
if (a == b) {
    value = "hello";
}
print(value); // The final variable 'newRowColumnChildren' can't be read because it is potentially unassigned at this point.  Ensure that it is assigned on necessary execution paths.
```

If the lint is smart enough to know `final` is wrong in the third case, can't it be smart enough to know `late` is unnecessary in the second example?

**Comment by bwilkerson** [comment]
> If the lint is smart enough to know final is wrong in the third case, can't it be smart enough to know late is unnecessary in the second example?

Possibly. But it would definitely want to be a diagnostic that users opt into because the code isn't wrong from the point of view of the specification.

@stereotype441 Does the definite assignment implementation have enough information to make it possible to catch a case like this? If not, how difficult would it be to gather the additional data? If so, is there currently any information retained after it's done running that a lint could look at, or would we need to run it again in order to get the information?

**Comment by stereotype441** [comment]
> @stereotype441 Does the definite assignment implementation have enough information to make it possible to catch a case like this? If not, how difficult would it be to gather the additional data? If so, is there currently any information retained after it's done running that a lint could look at, or would we need to run it again in order to get the information?

Yes, I believe flow analysis already has the necessary information.  All the analyzer would need to do is record the value of `FlowAnalysis.isAssigned` at the site of every variable read, to see if the variable is definitely assigned at that point in code execution.  Then, for each `late` variable, the linter could check whether the variable was definitely assigned at all read sites; if it was, then the `late` was unnecessary.

Note: we would need to be careful to make sure that for null-aware and compound assignments (e.g. `variable ??= value` and `variable += value`) the analyzer calls `FlowAnalysis.isAssigned` prior to visiting the RHS, since the read of the variable occurs before the RHS is evaluated.  Here's an example test case:
```dart
late int x;
if (...) x = 0;
x += (x = 1); // `late` is necessary because read of `x` happens before `1` is written to `x`
```

> ...it would definitely want to be a diagnostic that users opt into because the code isn't wrong from the point of view of the specification.

Actually I think it would be ok to make this a hint that's always enabled, because there wouldn't be any false positives.  Similar to how we don't allow the user to disable "unused private declaration" hint.  Of course, the user should still be able to ignore it on a case by case basis using `ignore` comments.

**Comment by srawlins** [comment]
Duplicate of https://github.com/dart-lang/sdk/issues/43372

**Comment by pq** [comment]
Implemented w/ #3052.
